### PR TITLE
Memory barrier fixes for thread shifting

### PR DESCRIPTION
**Pull request by jjhursey**
* Replaces PR #385
 * Port over the atomics code
 * Add memory barriers from PR #386 (which is based upon discussion in PR #384
 * Fixes Issue #347

### CODE REVIEW

**Comment by jjhursey** [comment]
I'm looking into the `make dist` failure that CI found.

File: src/server/pmix_server_get.c [hunk 1]
```diff
@@ -603,9 +607,7 @@ static void dmdx_cbfunc(pmix_status_t status,
                         "[%s:%d] queue dmdx reply for %s:%d",
                         __FILE__, __LINE__,
                         caddy->lcd->proc.nspace, caddy->lcd->proc.rank);
-    event_assign(&caddy->ev, pmix_globals.evbase, -1, EV_WRITE,
-                 _process_dmdx_reply, caddy);
+    PMIX_THREADSHIFT(caddy, _process_dmdx_reply);
     event_priority_set(&caddy->ev, 0);
```
Comment (rhc54) [inline src/server/pmix_server_get.c]: I think you missed this line - needs to be removed

**Comment by jjhursey** [comment]
@rhc54 Fixed. Thanks!

**Comment by jjhursey** [comment]
@jjhursey Don't forget to bring over https://github.com/pmix/pmix/pull/391

**Comment by jjhursey** [comment]
I just pushed an update:
 - Adjust copyrights on the files changed in the second commit.
 - Cherrypick over the change from PR #391

**Pull request merged**